### PR TITLE
MM-11505 Update permalink view

### PR DESCRIPTION
**Pull request by sudheerDev**
#### Summary
  * Use existing channel view for permalinks
  * Add a new URL route so channelView can be used for permalinks
  * Add new history toast
  * Highlight of permalink post should fadeout after 6 secs. 
  * new permalink url changes to channel url after 6 secs. 
  
One of the bigger changes in the code is to remove the view state of focussedPostId as having that and depending on that causes issues when we change url to channel view. When url changes to channel view it will clear the redux view state and hence the selectors for posts will also change from permalink posts to channel posts. Instead we will be depending on react state instead for permalink post id and use that for selectors and clear that only on channel change again.

 Initial idea was to drop the use of `/pl/postId` url from UI and use only the new url `/channel/postId` but this will cause us to to expose channelNames at times raising few security and privacy concerns. 
We will use the old url in the UI but redirect internally to new url to show the channelView instead. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-11505
https://mattermost.atlassian.net/browse/MM-19783
https://mattermost.atlassian.net/browse/MM-19782
https://mattermost.atlassian.net/browse/MM-20614

#### Screenshots
Screenshot 2020-02-04 at 9 58 08 PM

### CODE REVIEW

**Comment by esethna** [comment]
@sudheerDev can you help get the test server up?

**Comment by sudheerDev** [comment]
> @sudheerDev can you help get the test server up?

Done

**Comment by esethna** [comment]
WIP: ST is working on removing the permalink highlight before this get's tested again

**Comment by esethna** [comment]
@sudheerDev please help identify any cases of compatibility or other concerns to help QA with what they should be on the lookout for

**Comment by sudheerDev** [comment]
> @sudheerDev please help identify any cases of compatibility or other potential areas of concerns to help QA with what they should be on the lookout for issues.

I don't think there will be any compatibility issues as we are not changing the permalink url exposed in the UI. 

It needs general permalink testing with DM, GM and channel.

**Comment by sudheerDev** [comment]
> Looks good @sudheerDev! Can we make the following adjustments if possible?
> 
> 1. If the permalink is from the same channel does it have to reload the page or can it scroll up?
> 2. Can we increase the time to remove the highlight by ~50%?

1. Ya, it is a little hard to do. As discussed offline lets have it as a separate ticket. 
2. Done

**Comment by jgilliam17** [comment]
Thank you @sudheerDev 

Tested, looks good just have one question (see below)
- Verified new permalink view opens in existing main channel view (on refresh)
- Jump to link opens on main channel view
- Permalink post highlight fades after 6 seconds
- Verified url changes from `/channel/channelName/postId`  to `/channel/channelName` ( thanks Sudheer for clarifying 🙂)
- Verified History toast (small issue with date toast logged separately MM-22757 out of scope for this pr)
- Verified compatibility with older mobile app versions 

*Question*: Since url changes to channel url after user clicks to view permalink, it affects going back/forth in history using `cmd+[` and `cmd+]`
When navigating through history, user sees bottom of the channel instead of permalink view of the post on that channel (that was previously viewed). Is this acceptable? @esethna

**Comment by esethna** [comment]
Yes that's fine, thanks for pointing that out though.

**Comment by esethna** [comment]
@sudheerDev did you figure out the weird behaviour with the jump to new messages button in permalink view if you hit the button before the 6 sec timeout?

**Comment by sudheerDev** [comment]
> @sudheerDev did you figure out the weird behaviour with the jump to new messages button in permalink view if you hit the button before the 6 sec timeout?

I didn't get to it. i will check this soon. if worst case reviews happen before, i think we can raise it a ticket as known issue. Do you think it is a blocker?

**Comment by esethna** [comment]
Okay thanks, a follow-up ticket is fine

File: components/permalink_view/permalink_view.jsx [hunk 1]
```diff
@@ -103,44 +77,6 @@ export default class PermalinkView extends React.PureComponent {
...
-                        />
-                        <LocalizedIcon
-                            className='fa fa-arrow-down'
-                            title={{id: t('center_panel.recent.icon'), defaultMessage: 'Jump to recent messages Icon'}}
-                        />
-                    </Link>
-                </div>
-            </div>
-        );
+        return null;
```
Comment (Willyfrog) [inline components/permalink_view/permalink_view.jsx]: we might want to add 
```js
console.warn(`DEPRECATED: ${window.location} is using deprecated permalink view`};
```
or some other way to inform of deprecation, not sure we have an stablished one.
so we know if something is trying to use this and we didn't notice

File: components/channel_layout/channel_identifier_router/channel_identifier_router.jsx [hunk 2]
```diff
@@ -36,11 +40,27 @@ export default class ChannelIdentifierRouter extends React.PureComponent {
...
+    componentWillUnmount() {
+        clearTimeout(this.replaceUrlTimeout);
+    }
+
+    replaceUrlIfPermalink = () => {
+        if (this.props.match.params.postid) {
+            this.replaceUrlTimeout = setTimeout(() => {
+                const channelUrl = this.props.match.url.split('/').slice(0, -1).join('/');
+                browserHistory.replace(channelUrl);
+            }, Constants.PERMALINK_FADEOUT);
```
[the remark this answers is not in the document]
Comment (sudheerDev) [reply components/channel_layout/channel_identifier_router/channel_identifier_router.jsx]: Click something in the view? Triggering this change or url does not change the posts in center channel view. No visible difference other than removing highlight and changing url. 
if user clicks on something in LHS then the timer is cancelled in unMount life cycle

File: components/delete_post_modal/delete_post_modal.jsx [hunk 3]
```diff
@@ -34,16 +33,30 @@ export default class DeletePostModal extends React.PureComponent {
...
 
-        if (post.id === focusedPostId && channelName) {
-            browserHistory.push('/' + teamName + '/channels/' + channelName);
+        const matchUrlForDMGM = matchPath(this.props.location.pathname, {
+            path: '/:teamName/messages/:username/:postid',
+        });
+
+        const matchUrlForChannel = matchPath(this.props.location.pathname, {
+            path: '/:teamName/channels/:channelname/:postid',
+        });
```
[the remark this answers is not in the document]
Comment (sudheerDev) [reply components/delete_post_modal/delete_post_modal.jsx]: I am not sure that is a good option.

Anything with 4 slashes will match if we do that. `/anything/something/someName/ssup`. I feel having a fixed string in between makes it much more predictable and safe.

File: components/permalink_view/permalink_view.jsx (hunk 1 above)
Comment (sudheerDev) [reply components/permalink_view/permalink_view.jsx]: `null` here is to make sure we don't show anything in the UI when url is about to change from `/pl/postId` to `/channel/postId`. As of now we will have the old one as the default url in the view for user. We internally redirect so pretty much any  permalink url will trigger the log. 

Let me know if it isn't clear from the comment

File: components/delete_post_modal/delete_post_modal.jsx [hunk 4]
```diff
@@ -34,16 +33,30 @@ export default class DeletePostModal extends React.PureComponent {
...
+
+        if (matchUrlForDMGM) {
+            permalinkPostId = matchUrlForDMGM.params.postid;
+        } else if (matchUrlForChannel) {
+            permalinkPostId = matchUrlForChannel.params.postid;
+        }
```
[the remark this answers is not in the document]
Comment (sudheerDev) [reply components/delete_post_modal/delete_post_modal.jsx]: i tried something like this before but if there is no match then we are left with no object keys on `matchURL` so this throws an error as there is no `name` key. I added defensive check with ternary operator but it felt like a long condition

like `matchUrl && (matchURL.name =)` etc so resorted to this.

**Comment by esethna** [comment]
@deanwhillier when you have a chance please help review as I'd like this to soak for a few weeks before feature complete

File: components/delete_post_modal/delete_post_modal.jsx [hunk 5]
```diff
@@ -34,16 +33,30 @@ export default class DeletePostModal extends React.PureComponent {
...
         } = this.props;
 
+        let permalinkPostId = '';
+
         const result = await actions.deleteAndRemovePost(post);
 
-        if (post.id === focusedPostId && channelName) {
-            browserHistory.push('/' + teamName + '/channels/' + channelName);
+        const matchUrlForDMGM = matchPath(this.props.location.pathname, {
+            path: '/:teamName/messages/:username/:postid',
```
Comment (sudheerDev) [inline components/delete_post_modal/delete_post_modal.jsx]: Note: Change this to be based on a constant so we don't mess this up when someone changes the format of url's

Comment (sudheerDev) [reply components/delete_post_modal/delete_post_modal.jsx]: On second thought, i kept the constasts just in the same file. Constants file is huge and adding them there when not using them anywhere else as of now did not make a whole lot of sense

**Pull request merged**